src/treetensornetworks/opsum_to_ttn.jl

```diff
@@ -221,11 +233,354 @@ function svdTTN(
   return H
 end
 
+function qn_svdTTN(
+  os::OpSum{C}, sites::IndsNetwork{VT,<:Index}, root_vertex::VT; kwargs...
+)::TTN where {C,VT}
+  # Function barrier to improve type stability
+  ValType = ITensors.determineValType(terms(os))
+  return qn_svdTTN(ValType, os, sites, root_vertex; kwargs...)
+end
+
+function qn_svdTTN(
+  ValType::Type{<:Number},
+  os::OpSum{C},
+  sites::IndsNetwork{VT,<:Index},
+  root_vertex::VT;
+  kwargs...,
+)::TTN where {C,VT}
+  mindim::Int = get(kwargs, :mindim, 1)
+  maxdim::Int = get(kwargs, :maxdim, 10000)
+  cutoff::Float64 = get(kwargs, :cutoff, 1e-15)
+  #ValType = ITensors.determineValType(ITensors.terms(os))  #now included as argument in function signature
+
+  # check for qns on the site indices
+  #FIXME: this check for whether or not any of the siteindices has QNs is somewhat ugly
+  #FIXME: how are sites handled where some sites have QNs and some don't?
+
+  thishasqns = any(v -> hasqns(sites[v]), vertices(sites))
+
+  # traverse tree outwards from root vertex
+  vs = reverse(post_order_dfs_vertices(sites, root_vertex))                                 # store vertices in fixed ordering relative to root
+  es = reverse(reverse.(post_order_dfs_edges(sites, root_vertex)))                          # store edges in fixed ordering relative to root
+  # some things to keep track of
+  ranks = Dict(v => degree(sites, v) for v in vs)                                           # rank of every TTN tensor in network
+  Vs = Dict(e => Dict{QN,Matrix{ValType}}() for e in es)                                    # link isometries for SVD compression of TTN
+  #inmaps = Dict(e => Dict{Pair{Vector{Op},QN},Int}() for e in es)                                   
+  #outmaps = Dict(e => Dict{Pair{Vector{Op},QN},Int}() for e in es)                                   
+  inmaps = Dict{Pair{NamedGraphs.NamedEdge{VT},QN},Dict{Vector{Op},Int}}()                  # map from term in Hamiltonian to incoming channel index for every edge
```

Comment:
```suggestion
  inmaps = Dict{Pair{edgetype_sites,QN},Dict{Vector{Op},Int}}()                  # map from term in Hamiltonian to incoming channel index for every edge
```
and then define `edgetype_sites = edgetype(sites)` at the top of the function.

Reply:
(Maybe could be something shorter like `vtype` or something like that.)

Reply:
In the code we use `NamedGraphs.NamedEdge{VT}` except for in one place, we use only VT. Is there a way to extract the VT from `edgetype_sites` (which is `NamedGraphs.NamedEdge{VT}`)?

Reply:
It's this line `op_cache = Dict{Pair{String,VT},ITensor}()` --- `VT` should match the return type of `site(st)` where `st` is an element of `term` from the `OpSum`.
Can we just use Dict{Pair{String,Any},ITensor} here, to be flexible w.r.t. different edgetype_sites etc.?

Reply:
You should be able to use `vertextype(sites)` (which is defined in `NamedGraphs`: https://github.com/mtfishman/NamedGraphs.jl/blob/1b356a9cd31dca18e1762fdff30989bd042eebd2/src/Graphs/abstractgraph.jl#L55-L56). It's the same as `eltype` from `Graphs.jl` (https://juliagraphs.org/Graphs.jl/dev/core_functions/interface/#Base.eltype-Union%7BTuple%7BType%7B%3C:AbstractGraph%7BT%7D%7D%7D,%20Tuple%7BT%7D%7D%20where%20T) but I think `eltype` is a weird name for that functionality.